File: src/js/Body.ts
import Delta from './Delta';
import Point from './Point';

export default class Body implements Point{
  acceleration: Delta;
  color: string;
  mass: number;
  radius: number;
  velocity: Delta;
  x: number;
  y: number;

  constructor({
    acceleration = new Delta,
    color,
    mass,
    radius,
    velocity = new Delta,
    x,
    y
  }: {
    acceleration?: Delta,
    color?: string,
    mass?: number,
    radius: number,
    velocity?: Delta,
    x: number,
    y: number
  }){
    this.acceleration = acceleration;
    this.radius = radius;
    this.color = color || '#' + Math.round(((this.radius - 15) / 10) * 255).toString(16) + '0000';
    // Area of a circle
    this.mass = mass || Math.PI * this.radius ** 2;
    this.velocity = velocity;
    this.x = x;
    this.y = y;
  }
}

File: src/cpp/nbody/body.cc
#include "body.h"

#include <string>

#include "vector.h"

namespace nbody {

Body::Body(float x, float y, float mass, float radius, std::string color,
        Vector velocity = Vector(), Vector acceleration = Vector())
    : Vector(x, y) {
  mass_ = mass;
  radius_ = radius;
  color_ = color;
  velocity_ = velocity;
  acceleration_ = acceleration;
}

} // namespace nbody
